fairscale/nn/wrap/auto_wrap.py

```diff
@@ -130,68 +169,85 @@ class ConfigAutoWrap:
         ConfigAutoWrap.kwargs = {}
-        ConfigAutoWrap.module_blocklist = []
+        ConfigAutoWrap.should_wrap = None
 
     def __enter__(self) -> None:
-        self.enable_autowrap_context(self.kwargs)
+        self.enable_autowrap_context(self.should_wrap, self.kwargs)
 
     def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
         self.disable_autowrap_context()
 
     @staticmethod
-    def recursive_wrap(module: nn.Module, min_num_params: int, **kwargs: Any) -> Tuple[nn.Module, int]:
+    def recursive_wrap(module: nn.Module, should_wrap: Optional[Callable], **kwargs: Any) -> Tuple[nn.Module, int]:
         """
         Automatically wrap child modules of *module* that meet the given
         criteria with :func:`auto_wrap`.
 
         Args:
-            module (nn.Module): module to recursively wrap
-            min_num_params (int): min number of parameters for a child Module to
-                be wrapped
+            module (nn.Module):
+                module to recursively wrap
+            should_wrap (Callable, Optional):
+                optionally, override the :func:`should_wrap` from the context.
+
+        Returns:
+            (nn.Module, int):
+                Wrapped module and the number parameters wrapped recursively.
         """
-        if isinstance(module, tuple(ConfigAutoWrap.module_blocklist)):
-            # If the module has been blocklisted from wrapping, we return
-            return module, 0
+        if should_wrap is None:
+            should_wrap = ConfigAutoWrap.should_wrap
+
+        # Make sure no child is not already wrapped.
+        for _, child in module.named_modules():
+            assert not isinstance(child, cast(type, ConfigAutoWrap.wrapper_cls))
 
+        # We count all params, assuming none of them is already wrapped.
         num_params = sum([p.numel() for p in module.parameters()])
 
-        if num_params >= min_num_params:
+        assert should_wrap is not None
+        if should_wrap(True, module, num_params):
             total_wrapped_params = 0
             # Iterate through the children, recursively wrap if necessary
             for name, child in module.named_children():
                 wrapped_child, num_wrapped_params = ConfigAutoWrap.recursive_wrap(
-                    module=child, min_num_params=min_num_params, **kwargs
+                    module=child, should_wrap=should_wrap, **kwargs
                 )
                 setattr(module, name, wrapped_child)
                 # Keep track of how many parameters have been wrapped
                 total_wrapped_params += num_wrapped_params
             # decide if we need to wrap the current module,
             # since the left over parameters exceed the number of params to wrap
             remainder = num_params - total_wrapped_params
-            if remainder >= min_num_params and not isinstance(module, tuple(FSDP_MODULE_EXCLUDE_WRAP)):
+            if should_wrap(False, module, remainder):
```

Comment:
same, kwargs would make this easier to read

Reply:
done